kernel/src/engine/sync/json.rs

```diff
@@ -52,7 +52,7 @@ impl JsonHandler for SyncJsonHandler {
     fn write_json_file(
         &self,
         path: &Url,
-        data: Box<dyn Iterator<Item = Box<dyn EngineData>> + Send>,
+        data: Box<dyn Iterator<Item = DeltaResult<Box<dyn EngineData>>> + Send + '_>,
```

Comment:
curious of what people think of this API change. if we don't propagate results to here then we end up in a weird situation of either having to traverse everything to filter out `Err` cases or just unwap stuff? is there a better way to handle?

Reply:
What are we trying to achieve? Seems like it'll still be an issue that we might write some data and then fail. Without checking the whole thing for `Err` cases I'm not sure how you can avoid that?

Reply:
We have things like expression evaluation happening at every chunk of the iterator (e.g. before we get here we do the transformation from 'write_metadata' to actions) which might fail. Thus we have to 'hold on' to those `Result`s and propagate them through to this API. And yea not sure if there's a way around that other than something that scans the whole thing ahead of time?

Reply:
> But are you going to somehow recover from those errors? Why not just propagate earlier?

probably not? but unsure how to propagate early without traversing the whole iterator.

> To make sure I'm getting it. This would for example propagate any errors in generate_commit_info down here, rather than early returning from commit() when generate_commit_info failed, right? Early fail in commit actually seems better, so I'm not seeing the benefit here.

since `generate_commit_info` just always appends the first batch and we could fail early I'll use `generate_adds` as an example: we lazily generate the adds to produce an `impl Iterator<Item = DeltaResult<Box<dyn EngineData>>>`. that is, `generate_adds` takes an iterator of `write_metadata` and produces an iterator of result<actions> (as engine data). The result is injected here since we do some fallible work like expression evaluation to transform into the actions. Then, we are left with handing this iterator of results to the write_json API unless we want to traverse the whole iterator prior to passing it off to check that one of the batches didn't fail.

Reply:
ahh, finally I get it. Yeah, we want to let the errors propagate until we need their result. This makes sense. Thanks for bearing with me :)